custom_components/ramses_cc/__init__.py

```diff
@@ -52,258 +90,337 @@
+                        ),
+                        vol.Required(ATTR_CODE): cv.matches_regex(r"^[0-9A-F]{4}$"),
+                        vol.Required(ATTR_PAYLOAD): cv.matches_regex(
+                            r"^[0-9A-F]{1,48}$"
+                        ),
+                    }
+                ),
+            )
 
-        self._entity_state_attrs = ()
+    async def async_save_client_state(self, _: dt | None = None) -> None:
+        """Save the client state to the application store."""
 
-        # NOTE: this is bad: self.update_ha_state(delay=5)
+        _LOGGER.info("Saving the client state cache (packets, schema)")
 
-    @property
-    def extra_state_attributes(self) -> dict[str, Any]:
-        """Return the integration-specific state attributes."""
-        attrs = {
-            a: getattr(self._device, a)
-            for a in self._entity_state_attrs
-            if hasattr(self._device, a)
+        (schema, packets) = self._client._get_state()
+        remote_commands = self._known_commands | {
+            k: v._commands for k, v in self._entities.items() if hasattr(v, "_commands")
         }
-        # TODO: use self._device._parent?
-        # attrs["controller_id"] = self._device.ctl.id if self._device.ctl else None
-        return attrs
 
-    async def async_added_to_hass(self) -> None:
-        """Run when entity about to be added to hass."""
-        self._broker._entities[self.unique_id] = self
-        async_dispatcher_connect(self.hass, DOMAIN, self.async_handle_dispatch)
+        await self._store.async_save(
+            {
+                "client_state": {"schema": schema, "packets": packets},
+                "remotes": remote_commands,
+            }
+        )
 
-    @callback  # TODO: WIP
-    def _call_client_api(self, func, *args, **kwargs) -> None:
-        """Wrap client APIs to make them threadsafe."""
-        # self.hass.loop.call_soon_threadsafe(
-        #     func(*args, **kwargs)
-        # )  # HACK: call_soon_threadsafe should not be needed
+    async def async_update(self, _: dt | None = None) -> None:
+        """Retrieve the latest state data from the client library."""
+        if not all(self._platforms.values()):
+            return
 
-        func(*args, **kwargs)
-        self.update_ha_state()
+        rf_entities: list[RamsesRFEntity] = []
+        for device in self._client.devices:
+            rf_entities.append(device)
+        for system in self._client.systems:
+            rf_entities.append(system)
+            if isinstance(system, MultiZone):
+                rf_entities.extend([z for z in system.zones if z.name])
+
+        have_new_entities = False
+        for rf_entity in rf_entities:
+            unique_id = f"{rf_entity.__class__.__module__}.{rf_entity.__class__.__qualname__}|{rf_entity.id}"
+            if unique_id in self._rf_entity_ids:
+                async_dispatcher_send(self._hass, f"RAMSES_RF_UPDATE_{rf_entity.id}")
```

Comment:
This uses an entity specific message as it may be nice in future to be able to check a last-updated timestamp here to limit the work done on each call. That would be a complicated endeavour though so for now we always trigger updates.